File: O-MI Node/src/main/scala/agentSystem/InputPusher.scala
package agentSystem

import database._
import parsing.Types._
import parsing.Types.Path._
import parsing.Types.OdfTypes._
import java.util.Date
import akka.actor._
import java.lang.Iterable
import scala.collection.JavaConversions.iterableAsScalaIterable
import scala.collection.JavaConversions.asJavaIterable

  case class HandleObjects(objs: Iterable[OdfObject])
  case class HandleInfoItems(items: Iterable[OdfInfoItem])
  case class HandlePathValuePairs(pairs: Iterable[(Path,OdfValue)])
  case class HandlePathMetaDataPairs(pairs: Iterable[(Path,String)])
/** Creates an object for pushing data from internal agents to db.
  *
  */
class DBPusher(val dbobject: DB) extends Actor with ActorLogging with IInputPusher{
 

  override def receive = {
    case HandleObjects(objs) => handleObjects(objs)
    case HandleInfoItems(items) => handleInfoItems(items)
    case HandlePathValuePairs(pairs) => handlePathValuePairs(pairs)
    case HandlePathMetaDataPairs(pairs) => handlePathMetaDataPairs(pairs)
  }
  /** Function for handling sequences of OdfObject.
    *
    */
  override def handleObjects( objs: Iterable[OdfObject] ) : Unit = {
    for(obj <- objs){
      if(obj.objects.nonEmpty)
        handleObjects(obj.objects)
      if(obj.infoItems.nonEmpty)
        handleInfoItems(obj.infoItems)
    }
    log.debug("Successfully saved Objects to DB")
  }

  /** Function for handling sequences of OdfInfoItem.
    *
    */
  override def handleInfoItems( infoitems: Iterable[OdfInfoItem]) : Unit = {
    
      val infos = infoitems.map{ info => 
        info.values.map{ tv => (info.path, tv) }
      }.flatten[(Path,OdfValue)].toList 
     dbobject.setMany(infos) 
    log.debug("Successfully saved InfoItems to DB")
  } 
  
  /** Function for handling sequences of path and value pairs.
    *
    */
  override def handlePathValuePairs( pairs: Iterable[(Path,OdfValue)] ) : Unit ={
    dbobject.setMany(pairs.toList)
    log.debug("Successfully saved Path-TimedValue pairs to DB")    
  }
  def handlePathMetaDataPairs( pairs: Iterable[(Path,String)] ): Unit ={
    pairs.foreach{ pair => dbobject.setMetaData(pair._1, pair._2)}
    log.debug("Successfully saved Path-MetaData pairs to DB")    
  }

}

File: O-MI Node/src/main/java/agentSystem/InputPusher.java
/**
  Copyright (c) 2015 Aalto University.

  Licensed under the 4-clause BSD (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at top most directory of project.

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
**/
package agentSystem;

import agentSystem.InputPusherCmds.HandleOdf;
import agentSystem.InputPusherCmds.HandleObjects;
import agentSystem.InputPusherCmds.HandleInfoItems;
import agentSystem.InputPusherCmds.HandlePathValuePairs;
import agentSystem.InputPusherCmds.HandlePathMetaDataPairs;
import types.OdfTypes.OdfObject;
import types.OdfTypes.OdfObjects;
import types.OdfTypes.OdfInfoItem;
import types.OdfTypes.OdfValue;
import types.Path;
import akka.actor.ActorRef;
import static akka.pattern.Patterns.ask;
import static akka.pattern.Patterns.pipe;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;
import akka.dispatch.Futures;
import akka.dispatch.Mapper;
import akka.util.Timeout;

import java.util.concurrent.TimeUnit;

import scala.Tuple2;

/** Interface for pushing data to InputPusher actor. 
 */
public class InputPusher {
    private static Timeout default_timeout = new Timeout(Duration.create(5, TimeUnit.SECONDS));
    public static ActorRef ipdb = null; 
     public static Future<Object> handleOdf( OdfObjects objs, Timeout t ) { 
    	if(ipdb != null)
        	return ask(ipdb,new HandleOdf(objs), t);
        else
            return Futures.failed(new Exception("ipdb is null for InputPusher."));
    }
     public static Future<Object> handleOdf( OdfObjects objs) { 
         return handleOdf( objs, default_timeout );
     }
    public static Future<Object> handleObjects( Iterable<OdfObject> objs, Timeout t ) { 
    	if(ipdb != null)
		return ask(ipdb,new HandleObjects(objs), t); 
        else
            return Futures.failed(new Exception("ipdb is null for InputPusher."));
    }
     public static Future<Object> handleObjects( Iterable<OdfObject> objs) { 
         return handleObjects( objs, default_timeout );
     }
    public static Future<Object> handleInfoItems( Iterable<OdfInfoItem> items,Timeout t ) { 
    	if(ipdb != null)
        	return ask(ipdb,new HandleInfoItems(items), t ); 
        else
            return Futures.failed(new Exception("ipdb is null for InputPusher."));
    }
     public static Future<Object> handleInfoItems( Iterable<OdfInfoItem> items) { 
         return handleInfoItems( items, default_timeout );
     }
    public static Future<Object> handlePathValuePairs(Iterable<Tuple2<Path,OdfValue>> pairs,Timeout t) { 
    	if(ipdb != null)
		return ask(ipdb,new HandlePathValuePairs(pairs),t);
        else
            return Futures.failed(new Exception("ipdb is null for InputPusher."));
    }
    public static Future<Object> handlePathValuePairs(Iterable<Tuple2<Path,OdfValue>> pairs) { 
         return handlePathValuePairs(pairs,default_timeout);
     }
    public static Future<Object> handlePathMetaDataPairs(Iterable< Tuple2<Path,String> > pairs, Timeout t) { 
    	if(ipdb != null)
	    return ask(ipdb,new HandlePathMetaDataPairs(pairs), t); 
        else
            return Futures.failed(new Exception("ipdb is null for InputPusher."));
    }
    public static Future<Object> handlePathMetaDataPairs(Iterable< Tuple2<Path,String> > pairs) { 
        return handlePathMetaDataPairs( pairs, default_timeout);
    } 
}
